### ISSUE TITLE
Performance regression in 2.11.2

### ISSUE THREAD

**Issue by cesar1000** [mockito/mockito#1330]
Hey,

We noticed a significant performance regression after upgrading mockito from 2.8.9 to 2.13.0. I ran a small test suite using the gradle profiler on all versions in maven from 2.8.9 to 2.15.0, and then I ran in all incremental versions between 2.11.0 and 2.12.0, and in 2.15.6. These are the results (running the same suite six times on each version):

```
* Running scenario lut2_8_9 using Gradle 4.5

Execution time 73718ms
Execution time 29881ms
Execution time 29732ms
Execution time 27989ms
Execution time 28106ms
Execution time 28022ms

* Running scenario lut2_8_47 using Gradle 4.5

Execution time 72891ms
Execution time 29856ms
Execution time 29565ms
Execution time 27953ms
Execution time 27860ms
Execution time 32560ms

* Running scenario lut2_9_0 using Gradle 4.5

Execution time 73294ms
Execution time 29473ms
Execution time 28919ms
Execution time 28123ms
Execution time 28774ms
Execution time 28520ms

* Running scenario lut2_10_0 using Gradle 4.5

Execution time 102599ms
Execution time 35798ms
Execution time 31482ms
Execution time 29246ms
Execution time 29455ms
Execution time 30323ms

* Running scenario lut2_11_0 using Gradle 4.5

Execution time 74937ms
Execution time 30467ms
Execution time 30107ms
Execution time 28431ms
Execution time 28475ms
Execution time 28725ms

* Running scenario lut2_11_1 using Gradle 4.5

Execution time 68621ms
Execution time 30310ms
Execution time 28658ms
Execution time 28737ms
Execution time 27995ms
Execution time 28019ms

* Running scenario lut2_11_2 using Gradle 4.5

Execution time 97055ms
Execution time 33968ms
Execution time 32878ms
Execution time 34173ms
Execution time 32074ms
Execution time 32275ms

* Running scenario lut2_11_3 using Gradle 4.5

Execution time 73529ms
Execution time 34654ms
Execution time 32788ms
Execution time 32006ms
Execution time 32662ms
Execution time 32262ms

* Running scenario lut2_11_4 using Gradle 4.5

Execution time 70969ms
Execution time 34695ms
Execution time 33143ms
Execution time 32760ms
Execution time 34144ms
Execution time 32873ms

* Running scenario lut2_11_5 using Gradle 4.5

Execution time 99748ms
Execution time 34378ms
Execution time 32722ms
Execution time 33579ms
Execution time 32017ms
Execution time 32096ms

* Running scenario lut2_11_6 using Gradle 4.5

Execution time 72502ms
Execution time 34446ms
Execution time 33008ms
Execution time 33511ms
Execution time 32114ms
Execution time 33485ms

* Running scenario lut2_11_7 using Gradle 4.5

Execution time 74509ms
Execution time 34985ms
Execution time 33280ms
Execution time 31924ms
Execution time 33993ms
Execution time 32788ms

* Running scenario lut2_12_0 using Gradle 4.5

Execution time 75779ms
Execution time 34807ms
Execution time 33389ms
Execution time 33419ms
Execution time 32927ms
Execution time 33291ms

* Running scenario lut2_13_0 using Gradle 4.5

Execution time 69006ms
Execution time 34419ms
Execution time 33316ms
Execution time 33070ms
Execution time 33873ms
Execution time 33435ms

* Running scenario lut2_15_0 using Gradle 4.5

Execution time 76587ms
Execution time 34738ms
Execution time 35150ms
Execution time 33306ms
Execution time 33262ms
Execution time 33922ms

* Running scenario lut2_15_6 using Gradle 4.5

Execution time 78360ms
Execution time 35069ms
Execution time 35042ms
Execution time 33083ms
Execution time 33035ms
Execution time 32422ms
```

Between 2.11.1 and 2.11.2 there's a performance regression of ~15% of the overall time, and we've noticed even larger discrepancies in larger test suites. Looking at the diff between both versions, the upgrade of byte-buddy from 1.7.4 to 1.7.7 stands out.

Then I took snapshots of the test runs under 2.11.1 and 2.11.2 in YourKit. Comparing both snapshots took to this commit in byte-buddy: https://github.com/raphw/byte-buddy/commit/ec77e3d576a44694d3ca37d4728a464f97abbf0d. The full time of the regression seems to be in `registeredFieldCacheEntries.values().contains(fieldCache)`, which probably makes sense, since the `contains()` call has linear cost.

@raphw, would you mind taking a look? Does this make sense?

**Comment by raphw** [comment]
Thanks for the detailed report. I got to do some testing myself snd the regression is caused by a bug fix for a rather obscure case. I did not anticipate that it would influence the happy path in this severe way, too. I will look into how this can be solved in a better way! Thanks for reporting.

**Comment by cesar1000** [comment]
Awesome, thanks! :)

**Comment by raphw** [comment]
@cesar1000 I did a small change. Could you try to build Byte Buddy from GitHub and run your performance tests against the latest version? I wonder if this already resolves the regression.

**Comment by cesar1000** [comment]
Ah, that works :). I ran a baseline without the fix, and then I replaced the dependency and tried again. The timings are back to what they used to be before 2.11.2:

```
* Running scenario lut2_15_0 using Gradle 4.5

Execution time 55164ms
Execution time 37855ms
Execution time 35356ms
Execution time 34410ms
Execution time 34937ms
Execution time 34755ms

* Running scenario lut2_15_0_fix using Gradle 4.5

Execution time 76607ms
Execution time 34037ms
Execution time 30709ms
Execution time 30919ms
Execution time 29251ms
Execution time 29324ms
```
Thanks for the fix!

Incidentally, I noticed that enabling inline mocks in this same test suite runs ~30% slower. Are inline mocks expected to be slower, or is that a sign of a performance issue?

**Comment by raphw** [comment]
Yes, inline mocks make heavy use of the instrumentation API what regularly triggers full GC and JIT deoptimizations. Unfortunately, this is a cost the backing technic implies.

Thank for double-checking. I am preparing a release of Byte Buddy this week that also supports Java 10 and 11 and the performance fix will come along with it.

**Comment by cesar1000** [comment]
Ah, thanks, that's good context. And thanks for the fix!

**Comment by TimvdLippe** [comment]
@cesar1000 Could you confirm this has been fixed with ByteBuddy 1.8.0 (shipped with Mockito 2.18.0)?

**Comment by cesar1000** [comment]
Thanks, @TimvdLippe, I ran my tests again in both 2.11.0 and 2.18.0 and found performance to be similar or even slightly better :)
